src/sentry/tasks/on_demand_metrics.py

```diff
@@ -330,44 +335,70 @@ def _get_widget_query_low_cardinality(
     New queries will be checked upon creation and not allowed at that time.
     """
 
-    max_cardinality_allowed = options.get("on_demand.max_widget_cardinality.count")
     cache_key = _get_widget_query_cardinality_cache_key(widget_query)
-
-    # We default low cardinality to true since if it's false we'll remove user data.
-    is_low_cardinality = cache.get(cache_key, default=True)
     query_columns = widget_query.columns
+    max_cardinality_allowed = options.get("on_demand.max_widget_cardinality.count")
+
+    return all(
+        check_field_cardinality(
+            query_columns, organization, max_cardinality_allowed, cache_key
+        ).values()
+    )
+
 
+def check_field_cardinality(
+    query_columns: list[str],
+    organization: Organization,
+    max_cardinality: int,
+    widget_cache_key: str | None = None,
+    widget_query: DashboardWidgetQuery | None = None,
+) -> dict[str, str]:
     if not query_columns:
         return None
 
+    # We cache each key individually to query less
```

Comment:
May want to add a rollout/killswitch here, just in case.

Reply:
I'll add the killswitch here: https://github.com/getsentry/sentry/pull/65051